Change `SwitchTarget` representation in StableMIR

celinval opened the pull request:
The new structure encodes its invariant, which reduces the likelihood of having an inconsistent representation. It is also more intuitive and user friendly.

I encapsulated the structure for now in case we decide to change it back.

### Notes:

1. I had to change the `Successors` type, since there's a conflict on the iterator type. We could potentially implement an iterator here, but I would prefer keeping it simple for now, and add a `successors_iter()` method if needed.
2. I removed `CoroutineDrop` for now since it we never create it. We can add it when we add support to other MIR stages.

celinval committed:
Change SwitchTarget representation

The new structure encodes its invariant, which reduces the likelihood
of having an inconsistent representation. It is also more intuitive and
user friendly.

I encapsulated the structure for now in case we decide to change it back.

ouz-a commented:
Thanks! This looks good, have you tested if we are correctly printing `otherwise` ?

celinval commented:
It looks correct. The following function:
```rust
```

prints the following:
```rust
fn demux(_0: u8) -> u8 {
}
    bb0: {
        switchInt(_1) -> [0: bb1, 1: bb2, 2: bb3, otherwise: bb4]
    }
    bb1: {
        _0 = const 0_u8
        goto -> 5
    }
    bb2: {
        _0 = const 10_u8
        goto -> 5
    }
    bb3: {
        _0 = const 6_u8
        goto -> 5
    }
    bb4: {
        _0 = const 8_u8
        goto -> 5
    }
    bb5: {
        return
    }
```

I believe the issues with this output are unrelated to this change, and you are addressing them in #118364, right?

ouz-a commented:
@bors r+ rollup

celinval commented:
After inspecting the output, I realized the basic blocks order is incorrect.

@bors r-

celinval committed:
Fix SwitchTarget pretty print

We currently rely on the order of successors to be conditional branches
first, followed by the otherwise target.

ouz-a commented:
@bors r+ rollup

bors commented:
:hourglass: Testing commit 9d2c92377d1bf4d8b408b57b681bb6a04eb153a6 with merge c263ccf18571b4a3791e4c50dcae5473d6f9593a...
<!-- homu: {"type":"BuildStarted","head_sha":"9d2c92377d1bf4d8b408b57b681bb6a04eb153a6","merge_sha":"c263ccf18571b4a3791e4c50dcae5473d6f9593a"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/7054806646/job/19206337027)
Approved by: ouz-a
Pushing c263ccf18571b4a3791e4c50dcae5473d6f9593a to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"ouz-a","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/7054806646/job/19206337027"},"merge_sha":"c263ccf18571b4a3791e4c50dcae5473d6f9593a"} -->

rust-timer commented:
Finished benchmarking commit (c263ccf18571b4a3791e4c50dcae5473d6f9593a): [comparison URL](https://perf.rust-lang.org/compare.html?start=1d726a2be0b9a762c5a15154e65f925994a58445&end=c263ccf18571b4a3791e4c50dcae5473d6f9593a&stat=instructions:u).

### Overall result: no relevant changes - no action needed


@rustbot label: -perf-regression



### [Instruction count](https://perf.rust-lang.org/compare.html?start=1d726a2be0b9a762c5a15154e65f925994a58445&end=c263ccf18571b4a3791e4c50dcae5473d6f9593a&stat=instructions:u)
This benchmark run did not return any relevant results for this metric.

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=1d726a2be0b9a762c5a15154e65f925994a58445&end=c263ccf18571b4a3791e4c50dcae5473d6f9593a&stat=max-rss)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 2.7%  | [2.7%, 2.7%]   | 1     |
| Regressions ❌ <br /> (secondary)  | 2.3%  | [2.3%, 2.3%]   | 1     |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -2.1% | [-2.1%, -2.1%] | 1     |
| All ❌✅ (primary)                 | 2.7%  | [2.7%, 2.7%]   | 1     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=1d726a2be0b9a762c5a15154e65f925994a58445&end=c263ccf18571b4a3791e4c50dcae5473d6f9593a&stat=cycles:u)
This benchmark run did not return any relevant results for this metric.

### [Binary size](https://perf.rust-lang.org/compare.html?start=1d726a2be0b9a762c5a15154e65f925994a58445&end=c263ccf18571b4a3791e4c50dcae5473d6f9593a&stat=size:linked_artifact)
This benchmark run did not return any relevant results for this metric.

**Bootstrap**: 673.779s -> 673.487s (-0.04%)
**Artifact size**: 313.43 MiB -> 313.39 MiB (-0.01%)
